tests/tests/th_quant.plf

```diff
@@ -0,0 +1,81 @@
+; Deps: smt.plf th_base.plf
+(declare forall (! s sort
+                   (! t (term s)
+                      (! f formula
+                         formula))))
+
+;This program recursively checks the instantiation.
+;Composite terms (such as "apply _ _ ...") are handled recursively.
+;Then, if ti and t are equal, we return true. Otherwise, we first verify that t is the variable for which ti is substitued (ifmarked). if this is the case, ti should be equal to k.
+
+(program is_inst_t ((ti term) (t term) (k term)) bool
```

Comment:
I actually was expecting that *only* the new style of side condition works.  It is possible to pass the type parameters and have `ti`, `t`, `k` have types `(term s1)`, `(term s1)`, `(term s2)` like we discussed?  Or do we want to continue supporting this version as well?

Reply:
My apologies --- I should have made this more clear in the PR description.

Ultimately I'd like to do what you suggest.

Specifically I'd like to dis-allow dependent functions as input/return types while still allowing side conditions like this to be written, by implementing dependent type-checking for side conditions and rewriting the side conditions as you suggest.

However, for that to work, first we need to allow all proper types to be the input/return types of side conditions, otherwise `(term s1)` wouldn't be permitted anyway.

The latter change can be implemented independently of the former. This PR implements the latter. The former will follow at some point, assuming that I can get it working.